### PR TITLE
Disable Pagination on Requests

### PR DESCRIPTION
**Pull request by johannesschobel**
This fix lets requestors disable pagination by simply adding `?limit=0` to the URL. Otherwise it is not possible to circumvent a paginated resultset, if the dev used `return $repository->paginate()`.

This can be handy, if a full-blown client with stable and broadband connection would like to fetch all data. Mobile clients, however, would still get a paginated result!

### CODE REVIEW

**Comment by twcarefoot** [comment]
This seems like an easy way to hammer an application with a large data set. 

Disabling pagination at a global level like this defeats the purpose of implementing it in the first place. 

I completely agree that there are situations where you would want to return your entire data set. But why not just write a separate route or a specialized task in those cases?

Another solution to this could be to implement a function "paginateOrAll()" in the Repository class which 
 paginates by default and allows the "limit=0" switch.

**Comment by johannesschobel** [comment]
I don't like the "specific route" approach, as you would add redundancy to your application / action / task (wherever you add it).

hm, i think, the `paginateOrAll()` approach would be better, if @Mahmoudz would like it.
If you like, i could "revert" (i.e., make another PR) the PR and correct the behaviour tomorrow.. What do you think!?

**Comment by Mahmoudz** [comment]
@twcarefoot thanks for the clarification. @johannesschobel  yep I guess fixing this is essential, kindly submit another PR. I'll be reverting this PR now. Thanks.

**Comment by johannesschobel** [comment]
Hey @twcarefoot and @Mahmoudz ..
Actually, i don't know, if this is a problem.. Because, an attacker would also have the possibility to call `?limit=1000000000` in order to hammer the application.

Keep in mind, that you also have "route throttling" in order to limit the requests a specific client can make in a given timeperiod (e.g., 30 requests per minute).

Would you like me to create a dedicated `paginateOrAll()` method or leave it in the `paginate()` method and create a global `config()` flag to enable `unlimited pagination`?

**Comment by twcarefoot** [comment]
@johannesschobel 

> Actually, i don't know, if this is a problem.. Because, an attacker would also have the possibility to call ?limit=1000000000 in order to hammer the application.

Isn't that the purpose of the PAGINATION_LIMIT?

There might be a hybrid solution for this feature. https://github.com/apiato/apiato/issues/200  

Imagine each model could define it's own pagination limit and could allow pagination or to return all results.

For example; A world atlas project with 3 Models, `Country`, `Region`, `City`

`Country` could allow overriding the pagination since it makes sense to display all countries in the world in one call.

`Region` and `City` could define their own variable limit or just use the app default limit.

**Comment by johannesschobel** [comment]
the PAGINATION_LIMIT sets the default value how many elements should be returned..

I am a fan of giving the requester the possibility to decide on his own how many / what he would like to retrieve. It would be kind of "awkward" if he requests 100 elements, but the server would only return 20, because the limit is set to 20..

In my opinion, a config-flag `allow_disable_pagination` (default `false`) would be a good trade-off.. I wouldn't mind, defining this config flag in the `Repository`, so a dev could decide for each repository on its own (basically some kind of approach as you decided). But - at least for me - this should be done in the repository and **not** in the model, as the model only represents one element of the repository..

So basically, one could define
```php
protected $disablePagination = true;
```
in the `CountryRepository` (to pick up your example). This would give the requestor the the possibility to query with `?limit=0` in order to get all countries that fit his criteria (remember that he could also search!).
If `$disablePagination` is set to `false` a query with `?limit=0` would basically return the PAGINATION_LIMIT amount of entries (if as many are available).

What do you think about this?! I would be happy to implement it this way - that would be a good trade-off, i think!

**Comment by twcarefoot** [comment]
Yes I completely agree to add this property to the repository instead of the model.

Except it still doesn't allow each repository to have it's own custom limit;

A couple possible implementations to consider:
`protected $paginationLimit = 0;` - Makes sense with the `limit=0` use case and using integers
`protected $paginationLimit = false;` - This is explicitly saying there is no limit
`protected $paginationLimit = null;`  - Not my favourite, it's a bit wishy washy

And you could naturally set an integer like so
`protected $paginationLimit = 250;`

On a side note, PAGINATION_LIMIT is a bit of a misnomer. PAGINATION_LIMIT_DEFAULT would make more sense. Thanks for clearing that up

**Comment by johannesschobel** [comment]
i think, the requestor (i.e., the client application) should decide, how many items he would like to have on one page. If you (dev) would set the upper limit "hardcoded" (e.g., by defining `$paginationLimit = 25`) and a user would like to have 100 entries on one page, this would **not** be possible, because you would strip him down to 25 (instead of the desired 100) entries, which he would could handle..

I will implement a `$disablePagination` flag for the repositories that would make this behaviour configurable tomorrow. Furthermore, I will update the docs in order to clarify the purpose of `limits` and give an example..

Cheers
